Improve the SoA accessors and optimisations

fwyzard opened the pull request:
#### PR description:

These changes fix various issues with the SoA accessors:
  - return const scalars by const reference instead of by value, which would break non-trivially copiable types;
  - remove the `__builtin_assume_aligned` compiler hints from the SoA fields, as it breaks the `__restrict__` optimizations in nvcc;
  - add support for SoA view-level indexed access of Eigen columns;
  - add range checking to SoA view-level indexed accessors.

Extend the tests to check the access by (const) reference, the access to Eigen SoA columns, and the improved range checking.

#### PR validation:

Fixes a simple test provided by @borzari .

The updated unit test passes all checks.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

To be backported to CMSSW 12.5.x to ease the migration to Alpaka.

fwyzard committed:
Fix the return type for const view of scalars

Return const scalars by const reference instead of by value, which would
break non-trivially copiable types.

fwyzard committed:
Update the SoA tests to check the fields' address

Update the SoA tests to check the various ways of taking the address of
a column, scalar, or Eigen column fields.

fwyzard commented:
@ericcano could you double check these changes, and in particular
  - does the const view still generate accesses through the constant cache ?
  - does the syntax for the return types look correct to you ?
  - why does `view[0].m()` works and `view.m(0)` does not ?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39919/32834

- This PR adds an extra 28KB to repository

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

- DataFormats/SoATemplate (**heterogeneous**)
- HeterogeneousCore/AlpakaTest (**heterogeneous**)


@cmsbuild, @makortel, @fwyzard can you please review it and eventually sign? Thanks.
@makortel, @missirol, @rovere this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

ericcano committed:
Remove __builtin_assume_aligned compiler hints from SoA fields

Remove the __builtin_assume_aligned compiler hints from the SoA fields, as it breaks the __restrict__ optimizations in nvcc.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39919/33010

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #39919 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

ericcano committed:
Add support for SoA view-level indexed access of Eigen columns

Add support for SoA view-level indexed access of Eigen columns and enable the corresponding test.

fwyzard commented:
please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39919/33045

- This PR adds an extra 36KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39919/33045/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39919/33045/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented:
enable gpu

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39919/33046

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #39919 was updated. @makortel, @fwyzard can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d75cf1/29004/summary.html
**COMMIT**: 4f89eee8762aa44960de9c4211f4710decd628a2
**CMSSW**: CMSSW_12_6_X_2022-11-14-2300/el8_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39919/29004/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3417074
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3417049
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-14-2300+d75cf1/54092/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19876
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19867
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_6_X_2022-11-14-2300+d75cf1/54074/triggerResults): no differences found

ericcano committed:
Add range checking to SoA view-level indexed accessors

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39919/33068

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #39919 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

fwyzard commented:
please test

fwyzard commented:
+heterogeneous

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

fwyzard commented:
@perrotta @rappoccio sorry the delay; can we have this in CMSSW_12_6_0_pre5 ?

The SoA code is not used in production anywhere, but having this in the (pre)release would ease the work of the various developers.

fwyzard commented:
type bugfix

perrotta commented:
+1

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d75cf1/29048/summary.html
**COMMIT**: 067becf9bc2afa9257c83e4261d11714746ba786
**CMSSW**: CMSSW_12_6_X_2022-11-16-1100/el8_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39919/29048/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test test-das-selected-lumis had ERRORS
</pre>

## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 12 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3417074
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3417040
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-16-1100+d75cf1/54139/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19876
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19864
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_6_X_2022-11-16-1100+d75cf1/54133/triggerResults): no differences found

perrotta commented:
+1
- test-das-selected-lumis unit test error does not seem to be correlated: let check it further in tomorrow morning IB

perrotta commented:
merge
